exonum/src/messages2/authorization.rs

```diff
@@ -0,0 +1,153 @@
+use failure::Error;
+use hex::{FromHex, ToHex};
+
+use std::fmt;
+
+use super::EMPTY_SIGNED_MESSAGE_SIZE;
+use crypto::{
+    self, hash, Hash, PublicKey, SecretKey, Signature, PUBLIC_KEY_LENGTH, SIGNATURE_LENGTH,
+};
+
+/// `SignedMessage` can be constructed from a raw byte buffer which must have the following
+/// data layout:
+///
+/// | Position  | Stored data             |
+/// | - - - - - | - - - - - - - - - - - - |
+/// | `0..32`   | author's public key     |
+/// | `32`      | message class           |
+/// | `33`      | message type            |
+/// | `34..N`   | payload                 |
+/// | `N..N+64` | signature               |
+///
+/// `SignedMessage` will verify the size of the buffer and the signature provided in it.
+/// This allows to keep the raw message buffer, but avoid verifying its signature again
+/// as every `SignedMessage` instance is guaranteed to have a correct signature.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
+pub struct SignedMessage {
+    pub(in messages) raw: Vec<u8>,
+}
+
+impl SignedMessage {
+    /// Creates `SignedMessage` from parts.
+    pub(crate) fn new(
+        class: u8,
+        tag: u8,
+        value: Vec<u8>,
+        author: PublicKey,
+        secret_key: &SecretKey,
+    ) -> SignedMessage {
+        let mut buffer = Vec::with_capacity(2 + value.len() + PUBLIC_KEY_LENGTH + SIGNATURE_LENGTH);
```

Comment:
@vldm is it possible to get rid off the magic number ?

Reply:
Personally, I prefer to see something like that:
```rust
let mut buffer = Vec::with_capacity(mem::size_of_val(&class) + mem::size_of_val(&tag) + PUBLIC_KEY_LENGTH + SIGNATURE_LENGTH);
```